### DISCUSSION TITLE
Incoming Azure storage driver deprecation

### DISCUSSION THREAD

**Discussion by renchap** [mastodon/mastodon#31617] [General]
We are strongly considering deprecating the support for Azure Blob storage as a media backend in Mastodon 4.3 (currently targeted for release in September), and removing it entirely in Mastodon 4.4.

This feature was contributed by the community in Mastodon 4.2 (in https://github.com/mastodon/mastodon/pull/23607), but Azure stopped supporting their official SDK (which the feature is currently using). It means that there might be never-fixed security issues in them. In addition, they also rely on an old version the `faraday` gem, and this is causing dependency version conflict for Mastodon, preventing us to upgrade some critical gems.

Fixing this would require having a new Paperclip driver for Azure Blob storage not based on the formed Azure SDK, and none seem to exist at the moment. Someone created the [`azure-blob`](https://github.com/testdouble/azure-blob) gem as a much simpler replacement for the full Azure SDK, and this could be used to create a new Paperclip driver, but the current Mastodon team does not have the resources to do so (and we do not have any experience with Azure, nor an environment to properly test this).

The best solution here would be for the community to work on such Azure/Paperclip plugin so we can replace the existing implementation, but we do not see another path for us to continue supporting Azure as a media storage option.

It would help us to know if you are currently using Azure for media storage, to see how much of an impact this would have. We can probably provide some instructions to migrate from Azure Blog to an S3 backend (using `rclone`), if there is interest for it.
If someone wants to work on this Paperclip plugin, we might be able to provide guidance as well.

**Comment by mistydemeo** [comment]
I'm currently using this feature on digipres.club, and it would be of enormous use for me to be able to continue using it. I can review what it might take to migrate from the deprecated Azure SDK to another backend such as `azure-blob`.

Just as a note, since you mention not having the resources to work on this, it would have been helpful if you'd reached out to me in the first place since I wrote the original feature. I could easily have missed this question.

**Reply by renchap** [reply]
Hi @mistydemeo and sorry for the delay, I am just back from vacations.

I [pinged on the original PR about this deprecation](https://github.com/mastodon/mastodon/pull/23607#issuecomment-2311947401), but did not try to reach out to you directly.

We really need to move away from the retired Azure SDK library, and so far you are the only person that reached out to us saying that they use this feature, and we are a very small team, so we wont be able to reimplement this. Our current thinking is to mark this as deprecated in 4.3 (but not remove the feature), then once 4.3 is released, remove it from `main`, then if the community contributes an updated adapter, add it again and remove the deprecation.

If you want to work on this, we can help you to ensure this gets merged before 4.4 is released so we do not drop support for Azure storage. It would probably involve creating a new Paperclip adapter using `azure-blob`, similar to https://github.com/supportify/paperclip-azure, or update `paperclip-azure` to use `azure-blob`.

**Comment by jeremiahlee** [comment]
For reference: [_Retirement notice: The Azure Storage Ruby client libraries will be retired on 13 September 2024_](https://azure.microsoft.com/en-us/updates/retirement-notice-the-azure-storage-ruby-client-libraries-will-be-retired-on-13-september-2024/)

**Comment by JoeDupuis** [comment]
I missed this discussion. 
I have a PR up  to migrate paperclip-azure to azure-blob: https://github.com/mistydemeo/paperclip-azure/pull/4

**Reply by JoeDupuis** [reply]
CC @mistydemeo

**Reply by mistydemeo** [reply]
Thank you! Apologies, looks like I didn't get a notification on the PR. I'll review today or tomorrow.

**Reply by JoeDupuis** [reply]
No worries! Thank you!

**Reply by renchap** [reply]
That's awesome, thanks @JoeDupuis!

We plan to release Mastodon 4.3.0 RC1 next week, it would be great if we can update to the new `paperclip-azure` before then, so we can release it with the new gem.

**Reply by JoeDupuis** [reply]
I just opened the PR #32080
I did a couple tests and everything runs smoothly.
